File: nchito-android/app/src/main/java/com/owenphiri/nchito/data/NchitoApi.kt
package com.owenphiri.nchito.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * Thin PostgREST + RPC + Storage client over HttpURLConnection.
 *
 * A transport, not a data layer: it knows how to talk to Supabase and nothing
 * about gigs or wallets — that lives in LiveRepository. Mirrors the iOS
 * NchitoAPI so the two stay comparable.
 *
 * The caller's access token goes on every request, so Row Level Security applies
 * exactly as it does in the SQL. Without it PostgREST falls back to the anon
 * role and quietly returns nothing.
 */
class NchitoApi {

    class ApiException(message: String, val status: Int = -1) : Exception(message)

    private var accessToken: String = ""

    fun setAccessToken(token: String) { accessToken = token }

    companion object {
        val json = Json {
            ignoreUnknownKeys = true      // tolerate columns added by later migrations
            coerceInputValues = true
        }
        /** Zambian mobile data can be slow; fail in a human timeframe. */
        private const val TIMEOUT_MS = 20_000
    }

    suspend fun get(table: String, query: String): String =
        request("/rest/v1/$table?$query", "GET", null)

    suspend fun insert(table: String, body: String): String =
        request("/rest/v1/$table", "POST", body, prefer = "return=representation")

    /**
     * Calls a Postgres function. Most writes go through these rather than
     * direct table writes, because the rules — commission tiering, proof gates,
     * PIN checks — live in the functions and must not be skippable by a client.
     */
    suspend fun rpc(function: String, args: String = "{}"): String =
        request("/rest/v1/rpc/$function", "POST", args)

    suspend fun upload(bucket: String, path: String, bytes: ByteArray, contentType: String): String {
        requestRaw("/storage/v1/object/$bucket/$path", "POST", bytes, contentType)
        return path
    }

    private suspend fun request(
        path: String, method: String, body: String?, prefer: String? = null,
    ): String = requestRaw(path, method, body?.toByteArray(), "application/json", prefer)

    private suspend fun requestRaw(
        path: String, method: String, body: ByteArray?,
        contentType: String, prefer: String? = null,
    ): String = withContext(Dispatchers.IO) {
        if (!SupabaseConfig.isConfigured) throw ApiException("Supabase isn't configured yet.")
        if (accessToken.isEmpty()) throw ApiException("Please sign in again.")

        val connection = (URL(SupabaseConfig.PROJECT_URL + path).openConnection() as HttpURLConnection)
        try {
            connection.requestMethod = method
            connection.connectTimeout = TIMEOUT_MS
            connection.readTimeout = TIMEOUT_MS
            connection.setRequestProperty("apikey", SupabaseConfig.ANON_KEY)
            connection.setRequestProperty("Authorization", "Bearer $accessToken")
            connection.setRequestProperty("Content-Type", contentType)
            prefer?.let { connection.setRequestProperty("Prefer", it) }

            if (body != null) {
                connection.doOutput = true
                connection.outputStream.use { it.write(body) }
            }

            val status = connection.responseCode
            if (status !in 200..299) {
                val error = connection.errorStream
                    ?.bufferedReader()?.use(BufferedReader::readText).orEmpty()
                throw ApiException(friendlyMessage(status, error), status)
            }
            connection.inputStream.bufferedReader().use(BufferedReader::readText)
        } finally {
            connection.disconnect()
        }
    }

    /** 401/403 nearly always means an expired token or an RLS refusal, which the user can act on. */
    private fun friendlyMessage(status: Int, body: String): String {
        if (status == 401 || status == 403) return "Your session has expired. Please sign in again."
        val parsed = runCatching {
            json.parseToJsonElement(body).jsonObject["message"]?.jsonPrimitive?.content
        }.getOrNull()
        return parsed ?: "Server error ($status)."
    }
}

/** RPCs that return a bare scalar come back as a JSON string or number. */
fun String.asScalarString(): String = runCatching {
    NchitoApi.json.parseToJsonElement(this).jsonPrimitive.content
}.getOrDefault(this.trim('"'))

fun JsonElement.stringOrNull(key: String): String? =
    runCatching { jsonObject[key]?.jsonPrimitive?.content }.getOrNull()

File: nchito-ios/Nchito/Services/NchitoAPI.swift
import Foundation

/// Thin PostgREST + RPC + Storage client over URLSession.
///
/// Deliberately dependency-free, matching the rest of the app. It is a
/// transport, not a data layer: it knows how to talk to Supabase and nothing
/// about gigs or wallets — that lives in `LiveRepository`.
///
/// The caller's access token is sent on every request, so Row Level Security
/// applies exactly as it does in the SQL. Without it PostgREST would fall back
/// to the anon role and quietly return nothing.
actor NchitoAPI {

    enum APIError: LocalizedError {
        case notConfigured
        case notSignedIn
        case http(status: Int, message: String)
        case decoding(String)

        var errorDescription: String? {
            switch self {
            case .notConfigured:
                return "Supabase isn't configured yet."
            case .notSignedIn:
                return "Please sign in again."
            case .http(let status, let message):
                // 401/403 nearly always means the token expired or RLS refused,
                // which the user can act on; other codes they cannot.
                if status == 401 || status == 403 { return "Your session has expired. Please sign in again." }
                return message.isEmpty ? "Server error (\(status))." : message
            case .decoding(let detail):
                return "Unexpected response from the server. \(detail)"
            }
        }
    }

    private var accessToken: String = ""

    func setAccessToken(_ token: String) { accessToken = token }

    // MARK: - Requests

    /// A PostgREST table read. `query` is appended verbatim, e.g.
    /// "select=*&status=eq.open&order=created_at.desc&limit=20".
    func get<T: Decodable>(_ table: String, query: String, as: T.Type = T.self) async throws -> T {
        try await send(path: "/rest/v1/\(table)?\(query)", method: "GET", body: nil)
    }

    @discardableResult
    func insert<T: Decodable>(_ table: String, body: Encodable, as: T.Type = T.self) async throws -> T {
        try await send(path: "/rest/v1/\(table)", method: "POST",
                       body: try JSONEncoder.nchito.encode(AnyEncodable(body)),
                       prefer: "return=representation")
    }

    /// Updates rows in place. Used only for fields a user is allowed to change
    /// on their own row — their language, their chilimba auto-contribute
    /// setting — where row level security is the whole rule and there is no
    /// function's worth of logic to enforce.
    @discardableResult
    func update<T: Decodable>(_ table: String, query: String, body: Encodable,
                              as: T.Type = T.self) async throws -> T {
        try await send(path: "/rest/v1/\(table)?\(query)", method: "PATCH",
                       body: try JSONEncoder.nchito.encode(AnyEncodable(body)),
                       prefer: "return=representation")
    }

    /// Calls a Postgres function. Most of Nchito's writes go through these
    /// rather than direct table writes, because the rules (commission tiers,
    /// proof gates, PIN checks) live in the functions.
    @discardableResult
    func rpc<T: Decodable>(_ function: String, args: [String: Encodable] = [:],
                           as: T.Type = T.self) async throws -> T {
        let encoded = args.mapValues(AnyEncodable.init)
        return try await send(path: "/rest/v1/rpc/\(function)", method: "POST",
                              body: try JSONEncoder.nchito.encode(encoded))
    }

    /// Uploads bytes to a Storage bucket and returns the object path.
    func upload(bucket: String, path: String, data: Data, contentType: String) async throws -> String {
        _ = try await sendRaw(path: "/storage/v1/object/\(bucket)/\(path)",
                              method: "POST", body: data, contentType: contentType)
        return path
    }

    // MARK: - Plumbing

    private func send<T: Decodable>(path: String, method: String,
                                    body: Data?, prefer: String? = nil) async throws -> T {
        let data = try await sendRaw(path: path, method: method, body: body,
                                     contentType: "application/json", prefer: prefer)

        // Functions returning void or a bare scalar can come back empty.
        if data.isEmpty, let empty = EmptyResponse() as? T { return empty }

        do {
            return try JSONDecoder.nchito.decode(T.self, from: data)
        } catch {
            throw APIError.decoding(String(describing: error))
        }
    }

    private func sendRaw(path: String, method: String, body: Data?,
                         contentType: String, prefer: String? = nil) async throws -> Data {
        guard SupabaseConfig.isConfigured,
              let url = URL(string: SupabaseConfig.projectURL + path) else {
            throw APIError.notConfigured
        }
        guard !accessToken.isEmpty else { throw APIError.notSignedIn }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = body
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        if let prefer { request.setValue(prefer, forHTTPHeaderField: "Prefer") }
        // Zambian mobile data can be slow; fail in a human timeframe rather than
        // leaving a spinner up for a minute.
        request.timeoutInterval = 20

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError.http(status: -1, message: "No response")
        }
        guard (200..<300).contains(http.statusCode) else {
            // PostgREST reports problems as {"message": "...", "hint": "..."}.
            struct ServerError: Decodable { let message: String? }
            let parsed = try? JSONDecoder().decode(ServerError.self, from: data)
            throw APIError.http(status: http.statusCode, message: parsed?.message ?? "")
        }
        return data
    }
}

/// Lets a heterogeneous dictionary of arguments be encoded.
struct AnyEncodable: Encodable {
    private let encodeTo: (Encoder) throws -> Void
    init(_ wrapped: Encodable) { encodeTo = wrapped.encode }
    func encode(to encoder: Encoder) throws { try encodeTo(encoder) }
}

/// Stands in for a response body that carries nothing.
struct EmptyResponse: Codable {}
